modules/sensor_data.py

```diff
@@ -0,0 +1,223 @@
+        timestamp = entry[1]
+        if timestamp.month != date.month:
+            continue
+        if timestamp.year != date.year:
+            continue
+        filtered_data.update({entry[0]: format_entry(entry)})
+    return filtered_data
+
+
+def filter_year(sensor_data: list, date: Date) -> dict:
+    '''
+    Returns all entries matching year.
+    '''
+    filtered_data = {}
+    for entry in sensor_data:
+        timestamp = entry[1]
+        if timestamp.year != date.year:
+            continue
+        filtered_data.update({entry[0]: format_entry(entry)})
+    return filtered_data
+
+
+def format_entry(entry: list) -> dict:
+    '''
+    Beautifies the database entry.
+    sds011 has an entry length of 10.
+    dht22 has an entry length of 6.
+    '''
+    if len(entry) == 10:
+        return format_sds011(entry)
+    if len(entry) == 6:
+        return format_dht22(entry)
+    raise NotImplementedError()
+
+
+def format_sds011(entry: list) -> dict:
+    '''
+    Beatify database entry.
+    '''
+    return {
+        'timestamp': entry[1],
+        'P1': entry[2],
+        'durP1': entry[3],
+        'ratioP1': entry[4],
+        'P2': entry[5],
+        'durP2': entry[6],
+        'ratioP2': entry[7]
+    }
+
+
+def format_dht22(entry: list) -> dict:
+    '''
+    Beatify database entry.
+    '''
+    return {
+        'timestamp': entry[1],
+        'temperature': entry[2],
+        'humidity': entry[3]
+    }
+
+
+def get_sensor_function(sensor_type: str) -> Callable:
+    '''
+    Returns a function callable based on type if type is valid.
+    '''
+    match sensor_type:
+        case 'sds011':
+            insert_function = add_sds011_data
+        case 'dht22':
+            insert_function = add_dht22_data
+        case _:
+            raise HTTPException(detail=f'Unknown sensor type: {sensor_type}',
+                                status_code=status.HTTP_400_BAD_REQUEST)
+    return insert_function
+
+
+def add_sds011_data(con: Connection, file_data: list[list[str]]) -> None:
+    '''
+    Inserts into ParticulateMatterData due to sensor type being sds011.
+    '''
+    cur = con.cursor()
+    for line in file_data:
+        data = Sds011Data(timestamp=line[5],
+                          p1=line[6] if line[6] else 0,
+                          dur_p1=line[7] if line[7] else 0,
+                          ratio_p1=line[8] if line[8] else 0,
+                          p2=line[9] if line[9] else 0,
+                          dur_p2=line[10] if line[10] else 0,
+                          ratio_p2=line[11] if line[11] else 0,
+                          sensor_info=SensorInfo(id=line[0],
+                                                 type=line[1].lower())
+                          )
+        insert_sensor_data(cur, data)
+    cur.close()
+
+
+def add_dht22_data(con: Connection, file_data: list[list[str]]) -> None:
+    '''
+    Inserts into WeatherData due to sensor type being dht22.
+    '''
+    cur = con.cursor()
```

Comment:
Nit, wollen wir den cursor innerhalb dieser Funktion öffnen und nicht doch erst in insert_sensor_data weil dort die operation überhaupt erst getätigt wird? Außerdem managet man die Ressource dann nur innerhalb einer Funktion was es leichter macht zu maintainen falls man doch mehrere Sensoren hinzufügen sollte (Was nie passieren wir wegen Schulprojekt)